File: dart/Controladores_bases/ACbase_usuario.dart
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import 'package:crypto/crypto.dart';
import 'dart:convert';

class UserDatabaseController {
  static Database? _database;

  Future<Database> get database async {
    if (_database != null) return _database!;
    _database = await _initDatabase();
    return _database!;
  }

  Future<Database> _initDatabase() async {
    String path = join(await getDatabasesPath(), 'usuarios.db');
    return await openDatabase(path, version: 1, onCreate: _onCreate);
  }

  Future<void> _onCreate(Database db, int version) async {
    await db.execute('''
      CREATE TABLE IF NOT EXISTS users (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        username TEXT NOT NULL UNIQUE,
        password TEXT NOT NULL
      )
    ''');
  }

  String _hashPassword(String password) {
    var bytes = utf8.encode(password);
    var digest = sha256.convert(bytes);
    return digest.toString();
  }

  Future<bool> registrarUsuario(String username, String password) async {
    final db = await database;
    String hashedPassword = _hashPassword(password);
    try {
      await db.insert('users', {
        'username': username,
        'password': hashedPassword
      });
      return true;
    } on DatabaseException catch (e) {
      if (e.isUniqueConstraintError()) {
        return false;
      }
      rethrow;
    }
  }

  Future<bool> verificarLogin(String username, String password) async {
    final db = await database;
    String hashedPassword = _hashPassword(password);
    List<Map> result = await db.query('users',
        where: 'username = ? AND password = ?',
        whereArgs: [username, hashedPassword]);
    return result.isNotEmpty;
  }

  Future<Map<String, dynamic>?> consultarUsuario(String username) async {
    final db = await database;
    List<Map> result = await db.query('users',
        columns: ['id', 'username'],
        where: 'username = ?',
        whereArgs: [username]);
    if (result.isNotEmpty) {
      return result.first;
    }
    return null;
  }

  Future<bool> actualizarUsuario(String username, String newPassword) async {
    final db = await database;
    String hashedPassword = _hashPassword(newPassword);
    int count = await db.update('users', {'password': hashedPassword},
        where: 'username = ?', whereArgs: [username]);
    return count > 0;
  }

  Future<bool> eliminarUsuario(String username) async {
    final db = await database;
    int count = await db.delete('users', where: 'username = ?', whereArgs: [username]);
    return count > 0;
  }

  Future<List<Map<String, dynamic>>> listarUsuarios() async {
    final db = await database;
    return await db.query('users', columns: ['id', 'username']);
  }

  Future<void> cerrarCBaseUsuarios() async {
    final db = await database;
    await db.close();
  }
}

File: Python/Controladores_bases/ACbase_usuario.py
import sqlite3
from passlib.hash import pbkdf2_sha256

class base_usuario:
    def __init__(self, db_name="usuarios.db"):
        # aqui se conecta la base de datos con python
        self.conn = sqlite3.connect(db_name, check_same_thread=False)
        self.cursor = self.conn.cursor()
        self.crear_tabla()

    def crear_tabla(self):
        # aqui se crean las tablas necesarias
        self.cursor.execute('''CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL UNIQUE,
            password TEXT NOT NULL
        )''')
        self.conn.commit()

    def registrar_usuario(self, username, password):
        # aquí se crea el usuario con pass hasheado usando pbkdf2_sha256 para seguridad
        hashed = pbkdf2_sha256.hash(password)
        try:
            self.cursor.execute("INSERT INTO users (username, password) VALUES (?, ?)", (username, hashed))
            self.conn.commit()
            return True
        except sqlite3.IntegrityError:
            return False

    def verificar_login(self, username, password):
        # función para verificar el login
        self.cursor.execute("SELECT password FROM users WHERE username=?", (username,))
        result = self.cursor.fetchone()
        if result:
            stored_password = result[0]
            return pbkdf2_sha256.verify(password, stored_password)
        return False

    def Consultar_usuario(self, username):
        # función para consulta
        self.cursor.execute("SELECT id, username FROM users WHERE username=?", (username,))
        return self.cursor.fetchone()

    def actualizar_usuario(self, username, new_password):
        # función para editar la clave
        hashed = pbkdf2_sha256.hash(new_password)
        self.cursor.execute("UPDATE users SET password=? WHERE username=?", (hashed, username))
        self.conn.commit()
        return self.cursor.rowcount > 0

    def eliminar_usuario(self, username):
        # función para eliminar usuarios
        self.cursor.execute("DELETE FROM users WHERE username=?", (username,))
        self.conn.commit()
        return self.cursor.rowcount > 0

    def listar_usuarios(self):
        # listar todos los usuarios
        self.cursor.execute("SELECT id, username FROM users")
        return self.cursor.fetchall()

    def cerrar_c_base_usuarios(self):
        # cerrar la conexión a la base de datos
        self.conn.close()

